trin-metrics/src/utils.rs

```diff
@@ -0,0 +1,135 @@
+    observed: bool,
+    /// Starting instant for the timer.
+    start: Instant,
+}
+
+impl CustomHistogramTimer {
+    pub fn new(histogram: Histogram) -> Self {
+        Self {
+            histogram,
+            observed: false,
+            start: Instant::now(),
+        }
+    }
+
+    /// Observe and record timer duration (in seconds).
+    ///
+    /// It observes the floating-point number of seconds elapsed since the timer
+    /// started, and it records that value to the attached histogram.
+    pub fn observe_duration(self) {
+        self.stop_and_record();
+    }
+
+    /// Observe, record and return timer duration (in seconds).
+    ///
+    /// It observes and returns a floating-point number for seconds elapsed since
+    /// the timer started, recording that value to the attached histogram.
+    pub fn stop_and_record(self) -> f64 {
+        let mut timer = self;
+        timer.observe(true)
+    }
+
+    /// Observe and return timer duration (in seconds).
+    ///
+    /// It returns a floating-point number of seconds elapsed since the timer started,
+    /// without recording to any histogram.
+    pub fn stop_and_discard(self) -> f64 {
+        let mut timer = self;
+        timer.observe(false)
+    }
+
+    fn observe(&mut self, record: bool) -> f64 {
+        let v = Instant::now().saturating_duration_since(self.start);
+        let nanos = f64::from(v.subsec_nanos()) / 1e9;
+        let v = v.as_secs() as f64 + nanos;
+        self.observed = true;
+        if record {
+            self.histogram.observe(v);
+        }
+        v
+    }
+}
+
+impl Drop for CustomHistogramTimer {
+    fn drop(&mut self) {
+        if !self.observed {
+            self.observe(false);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{f64::EPSILON, thread, time::Duration};
+
+    use prometheus_exporter::prometheus::{
+        core::Collector, Histogram, HistogramOpts, DEFAULT_BUCKETS,
+    };
+
+    use crate::utils::CustomHistogramTimer;
+
+    /// Test taken from https://docs.rs/prometheus/0.13.3/src/prometheus/histogram.rs.html#1217-1260
+    /// Modified to work with CustomHistogramTimer
+    #[test]
+    fn test_histogram() {
+        let opts = HistogramOpts::new("test1", "test help")
+            .const_label("a", "1")
+            .const_label("b", "2");
+        let histogram = Histogram::with_opts(opts).unwrap();
+        histogram.observe(1.0);
+
+        let timer = CustomHistogramTimer::new(histogram.clone());
+        thread::sleep(Duration::from_millis(100));
+        timer.observe_duration();
+
+        // In this thread the timer isn't manually closed so its result will be dropped
+        let timer = CustomHistogramTimer::new(histogram.clone());
+        let handler = thread::spawn(move || {
+            let _timer = timer;
+            thread::sleep(Duration::from_millis(400));
+        });
+        assert!(handler.join().is_ok());
+
+        let mut mfs = histogram.collect();
```

Comment:
Hahah, all I can think of here is `motherf***ers`... Maybe that's just me? Imo it's worthwhile to update it to `metric_families` and the following `mf` -> `metric_family`, `m` -> `metric`... It makes the code more readable and avoids single letter variables